layout/plan.h

```diff
@@ -0,0 +1,619 @@
+	 * the entire range of data is available.
+	 *
+	 * This plop is useful in situations when the entire input data-set is
+	 * not available at once.
+	 *
+	 * @see m0_layout_inout_plop
+	 */
+	M0_LAT_IN_WRITE,
+	/**
+	 * Read plop, m0_layout_io_plop.
+	 *
+	 * This plop instructs the user to read data from the specified object.
+	 *
+	 * @see m0_layout_io_plop
+	 */
+	M0_LAT_READ,
+	/**
+	 * Write plop, m0_layout_io_plop.
+	 *
+	 * This plop instructs the user to write data to the specified object.
+	 *
+	 * @see m0_layout_io_plop
+	 */
+	M0_LAT_WRITE,
+	/**
+	 * Get plop, m0_layout_index_plop.
+	 *
+	 * This plop instructs the user to look values up in the specified
+	 * index.
+	 *
+	 * @see m0_layout_index_plop
+	 */
+	M0_LAT_GET,
+	/**
+	 * Put plop, m0_layout_index_plop.
+	 *
+	 * This plop instructs the user to insert values into the specified
+	 * index.
+	 *
+	 * @see m0_layout_index_plop
+	 */
+	M0_LAT_PUT,
+	/**
+	 * Fun plop, m0_layout_fun_plop.
+	 *
+	 * This plop instructs the user to execute a non-blocking function
+	 * provided by the implementation.
+	 *
+	 * @see m0_layout_fun_plop
+	 */
+	M0_LAT_FUN,
+	/**
+	 * Output plop, m0_layout_inout_plop.
+	 *
+	 * This plop signals the user that a part of the read data is available.
+	 *
+	 * @note this plop can be returned by the plan for the unit even before
+	 * m0_layout_plop_done() is called for its M0_LAT_READ plop. It is user
+	 * responsibility to track the dependencies between plops
+	 * (see m0_layout_plop::pl_deps), execute and call m0_layout_plop_done()
+	 * on them in order.
+	 *
+	 * @see m0_layout_inout_plop
+	 */
+	M0_LAT_OUT_READ,
+	/**
+	 * Done plop, m0_layout_plop.
+	 *
+	 * This plop signals the final completion of the plan execution. The
+	 * value of m0_layout_plop::pl_rc field of this plop contains the return
+	 * code of the entire operation.
+	 *
+	 * @see m0_layout_plop
+	 */
+	M0_LAT_DONE,
+
+	M0_LAT_NR
+};
+
+/**
+ * States of plops.
+ */
+enum m0_layout_plop_state {
+	M0_LPS_INIT,
+	M0_LPS_STARTED,
+	M0_LPS_DONE,
+};
+
+/**
+ * Common plop structure, shared by all "sub-classes".
+ *
+ * A call to m0_layout_plop_get() returns a structure starting with
+ * m0_layout_plop. The user should use m0_layout_plop::pl_type to determine
+ * actual sub-class.
+ */
+struct m0_layout_plop {
+	/** Plop type. */
+	enum m0_layout_plop_type         pl_type;
+	/** Plop state. */
+	enum m0_layout_plop_state        pl_state;
+	/** Plan this plop is part of. */
+	struct m0_layout_plan           *pl_plan;
+	/**
+	 * Plop colour.
+	 *
+	 * The implementation assigns colours to plops to improve locality of
+	 * reference. The caller can optionally map colours to processor cores
+	 * via m0_locality_get() and execute plops on cores corresponding to
+	 * their colours.
+	 */
+	uint64_t                         pl_colour;
+	/**
+	 * Linkage in the list of all plops in the plan.
+	 */
+	struct m0_tlink                  pl_all_link;
```

Comment:
Please clarify the link list head which the link is linked to.

Reply:
It's probably in the `m0_layout_plan` structure which will be introduced later during the implementation. Probably in a `plan_internal.h` file. I will update it then.